### PR TITLE
Add NoImpersonationAttribute so that special routes can deny access during impersonation

### PR DESCRIPTION
**Pull request by j3parker**

### CODE REVIEW

File: test/D2L.Security.OAuth2.WebApi.IntegrationTests/Authorization/NoImpersonationAttributeTests.cs [hunk 1]
```diff
@@ -0,0 +1,42 @@
+﻿using System.Net;
+using System.Threading.Tasks;
+using D2L.Services;
+using NUnit.Framework;
+
+namespace D2L.Security.OAuth2.Authorization {
+	[TestFixture]
+	internal sealed class NoImpersonationAttributeTests {
```
Comment (neverendingqs) [inline test/D2L.Security.OAuth2.WebApi.IntegrationTests/Authorization/NoImpersonationAttributeTests.cs]: Minor: want a test for no auth token? Weird case, but the `NoImpersonationAttribute` should return true.

File: test/D2L.Security.OAuth2.WebApi.IntegrationTests/TestWebService/Controllers/AuthorizationAttributeTestsController.cs [hunk 2]
```diff
@@ -7,39 +7,48 @@ namespace D2L.Security.OAuth2.TestWebService.Controllers {
...
 		public void UnspecifiedPrincipalType() {
 			// Will crash because we forgot an [AllowFrom(...)]	
 		}
 
 		[HttpGet]
-		[RequireScope("foo","bar","baz")]
+		[RequireScope( "foo", "bar", "baz" )]
 		[Authentication( users: true )]
-		[Route("authorization/basic")]
+		[Route( "authorization/basic" )]
```
Comment (neverendingqs) [inline test/D2L.Security.OAuth2.WebApi.IntegrationTests/TestWebService/Controllers/AuthorizationAttributeTestsController.cs]: Minor: "basic" makes me think of basic auth, but don't have a better name off the top of my head.

File: test/D2L.Security.OAuth2.WebApi.IntegrationTests/TestWebService/Controllers/AuthorizationAttributeTestsController.cs [hunk 3]
```diff
@@ -7,39 +7,48 @@ namespace D2L.Security.OAuth2.TestWebService.Controllers {
...
 		public void Anonymous() {
-			
+
+		}
+
+		[HttpGet]
+		[NoImpersonation]
+		[NoRequiredScope]
+		[Authentication( users: true, services: true )]
+		[Route( "authorization/imp" )]
```
Comment (neverendingqs) [inline test/D2L.Security.OAuth2.WebApi.IntegrationTests/TestWebService/Controllers/AuthorizationAttributeTestsController.cs]: `authorization/imp` => `authorization/noimp`?

File: src/D2L.Security.OAuth2.WebApi/Authorization/NoImpersonationAttribute.cs [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+using System.Web.Http.Controllers;
+using D2L.Security.OAuth2.Principal;
+
+namespace D2L.Security.OAuth2.Authorization {
+	/// <summary>
+	/// Block requests where the access token indicates user impersonation
+	/// </summary>
+	[AttributeUsage( AttributeTargets.All, AllowMultiple = false )]
+	public sealed class NoImpersonationAttribute : OAuth2AuthorizeAttribute {
+		protected override uint Order => 2;
```
Comment (neverendingqs) [inline src/D2L.Security.OAuth2.WebApi/Authorization/NoImpersonationAttribute.cs]: RE: https://ci.appveyor.com/project/Brightspace/d2l-security-oauth2/build/job/k83jy2420h5wustg/messages

.net 4.6.1?

FYI auth service is still at 4.5 (not sure if it matters).

File: test/D2L.Security.OAuth2.WebApi.IntegrationTests/TestWebService/Controllers/AuthorizationAttributeTestsController.cs (hunk 3 above)
Comment (j3parker) [inline test/D2L.Security.OAuth2.WebApi.IntegrationTests/TestWebService/Controllers/AuthorizationAttributeTestsController.cs]: ?

File: test/D2L.Security.OAuth2.WebApi.IntegrationTests/Authorization/NoImpersonationAttributeTests.cs (hunk 1 above)
Comment (j3parker) [inline test/D2L.Security.OAuth2.WebApi.IntegrationTests/Authorization/NoImpersonationAttributeTests.cs]: Not really worried about that; this attribute doesn't make sense without an `[Authentication]` (which will make that `null` check in the attribute not a concern) so not really interested in validating any particular behavior. If a route is anonymous there isn't any harm we could do either way.

File: src/D2L.Security.OAuth2.WebApi/Authorization/NoImpersonationAttribute.cs (hunk 4 above)
Comment (j3parker) [inline src/D2L.Security.OAuth2.WebApi/Authorization/NoImpersonationAttribute.cs]: sigh, I'll poke at that too because C# 6 is the bees knees.

**Comment by omsmith** [comment]
Checks themselves seem good

File: test/D2L.Security.OAuth2.WebApi.IntegrationTests/TestWebService/Controllers/AuthorizationAttributeTestsController.cs (hunk 3 above)
Comment (neverendingqs) [inline test/D2L.Security.OAuth2.WebApi.IntegrationTests/TestWebService/Controllers/AuthorizationAttributeTestsController.cs]: This route doesn't allow impersonation, but `authorization/imp` might come off as the opposite.

File: test/D2L.Security.OAuth2.WebApi.IntegrationTests/TestWebService/Controllers/AuthorizationAttributeTestsController.cs (hunk 2 above)
Comment (j3parker) [inline test/D2L.Security.OAuth2.WebApi.IntegrationTests/TestWebService/Controllers/AuthorizationAttributeTestsController.cs]: I don't think it's a constructive use of time to worry about people trying to infer the meaning of these tests solely from the route strings. It's pretty clear that this is just a basic test.

This test wasn't part of this PR other than my editor changing the whitespace.

File: test/D2L.Security.OAuth2.WebApi.IntegrationTests/TestWebService/Controllers/AuthorizationAttributeTestsController.cs (hunk 3 above)
Comment (j3parker) [inline test/D2L.Security.OAuth2.WebApi.IntegrationTests/TestWebService/Controllers/AuthorizationAttributeTestsController.cs]: I'm having a hard time viewing this comment as a legitimate use of time. It's pretty clear what's going on here.

**Pull request merged**